pkg/storj/metainfo.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storj
+
+import (
+	"context"
+)
+
+// Metainfo represents a database for storing meta-info about objects
+type Metainfo interface {
+	// MetainfoLimits returns limits for this metainfo database
+	Limits() (MetainfoLimits, error)
+
+	// CreateBucket creates a new bucket with the specified information
+	CreateBucket(ctx context.Context, bucket Bucket) (Bucket, error)
+	// DeleteBucket deletes bucket
+	DeleteBucket(ctx context.Context, bucket string) error
+	// GetBucket gets bucket information
+	GetBucket(ctx context.Context, bucket string) (Bucket, error)
+
+	// GetObject returns information about an object
+	GetObject(ctx context.Context, bucket string, path Path) (Object, error)
+	// GetObjectStream returns interface for reading the object stream
+	GetObjectStream(ctx context.Context, bucket string, path Path) (ReadonlyStream, error)
+
+	// CreateObject creates an uploading object and returns an interface for uploading Object information
+	CreateObject(ctx context.Context, bucket string, path Path, info Object) (MutableObject, error)
+	// ModifyObject creates an interface for modifying an existing object
+	ModifyObject(ctx context.Context, bucket string, path Path, info Object) (MutableObject, error)
+	// DeleteObject deletes an object from database
+	DeleteObject(ctx context.Context, bucket string, path Path) error
+
+	// TODO: add things for continuing existing objects
+	ListObjects(cctx context.Context, bucket string, options ListOptions) (ObjectList, error)
+}
+
+// ListOptions lists objects
+type ListOptions struct {
+	Prefix    Path
+	First     Path // First is relative to Prefix, full path is Prefix + First
+	Delimiter rune
+	Recursive bool
+	Limit     int
+
+	// Token used for the next listing.
+	Token string
+}
+
+// ObjectList is a list of objects
+type ObjectList struct {
+	Bucket string
+	Prefix Path
+	Token  string
+	More   bool
+
+	// Items paths are realtive to Prefix
+	// To get the full path use list.Prefix + list.Items[0].Path
+	Items []Object
+}
+
+// MetainfoLimits lists limits specified for the Metainfo database
+type MetainfoLimits struct {
+	// ListLimit specifies the maximum amount of items that can be listed at a time.
+	ListLimit int64
+
+	// MaximumInlineSegment specifies the maximum inline segment that is allowed to be stored.
+	MaximumInlineSegmentSize int64
```

Comment:
What's this? How would this info help the client?

Reply:
Stream uploading needs this information to know whether a segment can be uploaded to the metainfo database or it must be uploaded to piecestore.

Reply:
Oh, I don't know why I thought this was the maximum _number_ of inline segments that can be stored.

Anyway, the question is still valid. Why do we have this here? This is the client side configuration, not a server side one, like the `ListLimit`.

Reply:
I think it's safer to not allow uploading segments with inline that are larger than some number. E.g. 1GB inline segment might end up causing problems.

Reply:
Now as you said it, I remembered we already have such checks: https://github.com/storj/storj/blob/master/pkg/pointerdb/config.go#L28-L29

Not only for maximum size allowed for inline segments, but also for minimum size allowed. We should have them both here.

And perhaps make it clearer in the doc comments that these are server-side limits that the client cannot exceed.

Reply:
Perhaps, in a similar fashion to avoid abuse, we should have `MinimumRemoteSegmentSize` too.

Reply:
Which should be a check applied for all remote segments, but the last one.

Reply:
It would protect about creating too small remote segments, which would abuse the network by having to work with too small remote pieces.

I think this is exactly what the current MinInlineSegmentSize is for. It's a little bit weird we name it like this. MinRemoteSegmentSize would have been a better name...